File: network/java/src/main/java/org/aiddl/util/grpc/GrpcFunction.java
package org.aiddl.util.grpc;

import org.aiddl.core.interfaces.Function;
import org.aiddl.core.parser.Parser;
import org.aiddl.core.representation.SymbolicTerm;
import org.aiddl.core.representation.Term;
import org.aiddl.network.AiddlGrpc.AiddlBlockingStub;
import org.aiddl.network.AiddlOuterClass.FunctionCallRequest;
import org.aiddl.network.AiddlOuterClass.FunctionCallResult;

import io.grpc.Channel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;

/**
 * Wrap function around grpc call.
 * 
 * @author Uwe Koeckemann
 *
 */
public class GrpcFunction implements Function {

	String ip;
	int port;
	SymbolicTerm function_uri;

	private final AiddlBlockingStub blockingStub;
	
	public GrpcFunction( String host, int port, SymbolicTerm function_uri ) {
		  this(ManagedChannelBuilder.forAddress(host, port).usePlaintext().build(), function_uri);
		}

	public GrpcFunction( Channel channel, SymbolicTerm function_uri ) {
		blockingStub = AiddlGrpc.newBlockingStub(channel);
		this.function_uri = function_uri;		
	}

	@Override
	public Term apply(Term args) {
		FunctionCallRequest call_args = FunctionCallRequest.newBuilder().
				setFunctionUri(this.function_uri.toString()).
				setArg(args.toString()).
				build();

		FunctionCallResult result;
		try {
			result = blockingStub.functionCall(call_args);
			Term r = Parser.ParseTerm(result.getResult());
			return r;
		} catch (StatusRuntimeException e) {
			System.err.println(e);
			System.exit(1);
		}
		return null;
	}
}

File: external/grpc/scala/src/main/scala/org/aiddl/external/grpc/scala/container/GrpcFunction.scala
package org.aiddl.external.grpc.scala.container

import io.grpc.ManagedChannelBuilder
import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.representation.{Sym, Term}
import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.parser.Parser
import org.aiddl.core.scala.util.StopWatch
import org.aiddl.external.grpc.container.ResultStatus.SUCCESS
import org.aiddl.external.grpc.container.{ContainerGrpc, FunctionCallRequest, FunctionCallResult}
import org.aiddl.external.grpc.scala.converter.Converter

class GrpcFunction(host: String, port: Int, uri: Sym, c: Container) extends Function {
  val converter = new Converter(c)
  val parser = new Parser(c)

  val channel = ManagedChannelBuilder
    .forAddress(host, port)
    .usePlaintext()
    .build

  val blockingStub = ContainerGrpc.blockingStub(channel)

  override def apply(x: Term): Term = {

    StopWatch.start("Convert input")
    val xPb = this.converter.aiddl2pb(x)
    StopWatch.stop("Convert input")

    val request = FunctionCallRequest()
      .withFunctionUri(uri.toString)
      .withArg(xPb)

    StopWatch.start("Calling stub")
    val result = blockingStub.functionCall(request)
    StopWatch.stop("Calling stub")

    val r = result match {
      case FunctionCallResult(SUCCESS, Some(pbTerm), _) => {
        StopWatch.start("Convert output")
        val resultAiddl = this.converter.pb2aiddl(pbTerm)
        StopWatch.stop("Convert output")
        resultAiddl
      }
      case FunctionCallResult(_, _, _) => Sym("NIL")
    }
    r
  }
}
